http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/HealthCheckedChannelPool.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal;
+
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.pool.ChannelPool;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.Promise;
+import io.netty.util.concurrent.ScheduledFuture;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An implementation of {@link ChannelPool} that validates the health of its connections.
+ *
+ * This wraps another {@code ChannelPool}, and verifies:
+ * <ol>
+ * <li>All connections acquired from the underlying channel pool are in the active state.</li>
+ * <li>All connections released into the underlying pool that are not active, are closed before they are released.</li>
+ * </ol>
+ *
+ * Acquisitions that fail due to an unhealthy underlying channel are retried until a healthy channel can be returned, or the
+ * {@link NettyConfiguration#connectionAcquireTimeoutMillis()} timeout is reached.
+ */
+public class HealthCheckedChannelPool implements ChannelPool {
+    private final EventLoopGroup eventLoopGroup;
+    private final int acquireTimeoutMillis;
+    private final ChannelPool delegate;
+
+    public HealthCheckedChannelPool(EventLoopGroup eventLoopGroup,
+                                    NettyConfiguration configuration,
+                                    ChannelPool delegate) {
+        this.eventLoopGroup = eventLoopGroup;
+        this.acquireTimeoutMillis = configuration.connectionAcquireTimeoutMillis();
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Future<Channel> acquire() {
+        return acquire(eventLoopGroup.next().newPromise());
+    }
+
+    @Override
+    public Future<Channel> acquire(Promise<Channel> resultFuture) {
+        // Schedule a task to time out this acquisition, in case we can't acquire a channel fast enough.
+        ScheduledFuture<?> timeoutFuture =
+                eventLoopGroup.schedule(() -> timeoutAcquire(resultFuture), acquireTimeoutMillis, TimeUnit.MILLISECONDS);
+
+        tryAcquire(resultFuture, timeoutFuture);
+        return resultFuture;
+    }
+
+    /**
+     * Time out the provided acquire future, if it hasn't already been completed.
+     */
+    private void timeoutAcquire(Promise<Channel> resultFuture) {
+        if (!resultFuture.isDone()) {
```

Comment:
Originally, I didn't want the overhead of allocating the IOException (filling in the stack trace is apparently pretty expensive), because I wasn't cancelling the timeout on success. Now it's not really as needed, since I'm cancelling the timeout task, so I don't mind removing it.